### PR TITLE
Feature notification controller route

### PR DESCRIPTION
**Pull request by sholabioku**
### What this PR does (required):

-- Create notification
-- Mark notification as read
-- Get all notifications
--Get unread notifications

### Screenshots / Videos (front-end only):

- Post at least one screenshot to show the feature in action. **Only required in front-end work**

### Any information needed to test this feature (required):

- Post what steps to take to be able to test this feature (eg. "sign up as a new user and upload a profile picture from the navbar")

### Any issues with the current functionality (optional):

- List any problems where you were not able to complete all tasks on ticket
- List any problems this PR may cause for the project or other active PRs
- Post a screenshot/video of the problem, along with detailed console outputs where applicable

### CODE REVIEW

File: server/controllers/notification.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    data: unreadNotifications,
+    count: unreadNotifications.length,
+  });
+});
+
+// @desc Mark notifications as read
+// @route PATCH /notifications/read/:notificationId
+// @access Private
+exports.markNotificationAsRead = asyncHandler(async (req, res, next) => {
+  const notification = await Profile.findById(req.params.notificationId);
```
Comment (ugur-akin) [inline server/controllers/notification.js:46]: Did you mean `Notification.findById(...`?

File: server/controllers/notification.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+  if (!notification) {
+    res.status(404);
+    throw new Error(`No notification with id ${req.params.notificationId}`);
+  }
```
Comment (ugur-akin) [inline server/controllers/notification.js:51]: 👍🏼

File: server/controllers/notification.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+  const unreadNotifications = await Notification.find({ read: { $eq: false } });
+
+  if (!unreadNotifications) {
+    res.status(400);
+    throw new Error("Bad request");
+  }
```
Comment (ugur-akin) [inline server/controllers/notification.js:58]: I don't think we need this block here

File: server/controllers/notification.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+  const markAsRead = await Notification.findByIdAndUpdate(
+    req.params.notificationId,
+    { read: req.body.read },
+    {
+      new: true,
+    }
+  );
```
Comment (ugur-akin) [inline server/controllers/notification.js:66]: A couple of issues here:
- We don't need to `find` the notification again as we already `find` it at `line 46`. We can simply update the document in plain js and use `.save()` on the document.
- If we are marking a notification as read, then we don't need to receive a `read` param in the request body. We can just set `read` to `true`.

**Review by ugur-akin — COMMENTED** [review]
Hey Lukman, good work with this 👍🏼 Notifications are looking good, just a few minor issues to polish. 

Couldn't help but notice there was quite a lot more than just notifications related changes in this PR, I only skimmed through the other changes and tried to focus on the changes regarding notifications. Is there a particular reason for this? Why do you think it happened, is it possible to separate these changes into their own relevant PRs?

In any case, if we could avoid this in the future it would be most appreciated 🙏🏼

File: client/src/context/useAuthContext.tsx [hunk 5]
```diff
@@ -21,25 +21,24 @@ export const AuthProvider: FunctionComponent = ({ children }): JSX.Element => {
...
-      setLoggedInUser(data.user);
-      history.push('/dashboard');
-    },
+  const updateLoginContext = useCallback((data: AuthApiDataSuccess) => {
+    setLoggedInUser(data.user);
+    history.push('/dashboard');
     // eslint-disable-next-line react-hooks/exhaustive-deps
-    [history],
-  );
+  }, []);
```
[the remark this answers is not in the document]
Comment (sholabioku) [reply client/src/context/useAuthContext.tsx:30]: Those issues that are not related to to notifications came when I updated my local repository with upstream one...

[the remark this answers is not in the document]
Comment (sholabioku) [reply client/src/context/useAuthContext.tsx:30]: @ugur-akin I did not work on this ticket, I wonder how this came up, Perhaps, when I updated my local repo with remote repo

File: package.json [hunk 6]
```diff
@@ -8,6 +8,9 @@
     "prepare": "husky install"
   },
   "dependencies": {
-    "cloudinary": "^1.27.1"
+    "@mui/lab": "^5.0.0-alpha.55",
+    "cloudinary": "^1.27.1",
+    "install": "^0.13.0",
+    "npm": "^8.1.4"
```
[the remark this answers is not in the document]
Comment (sholabioku) [reply package.json:14]: @ugur-akin, These are work done by @gagandeep95

File: server/controllers/notification.js (hunk 1 above)
Comment (sholabioku) [reply server/controllers/notification.js:46]: @ugur-akin Yes, I made mistake

**Comment by sholabioku** [comment]
@ugur-akin Kindly check the fixes for the feedbacks

File: server/validate.js [hunk 7]
```diff
@@ -28,5 +27,16 @@ exports.validateLogin = [
     if (!errors.isEmpty())
       return res.status(400).json({ errors: errors.array() });
     next();
-  }
+  },
+];
+
+exports.validateNotification = [
```
Comment (rajivtitus) [inline server/validate.js:33]: 👍🏽

File: server/controllers/notification.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  res
+    .status(200)
+    .json({ success: true, data: notifications, count: notifications.length });
+});
+
+// @desc Fetch all unread notifications
+// @route GET /notifications/unread
+// @access Private
+exports.getUnreadNotifications = asyncHandler(async (req, res, next) => {
+  const unreadNotifications = await Notification.find({ read: { $eq: false } });
```
Comment (rajivtitus) [inline server/controllers/notification.js:34]: Won't this query fetch all the unread notifications for every user? We only need to fetch it for the logged in user :)

File: server/controllers/notification.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+//@route POST /notifications
+//@ access Private
+exports.createNotification = asyncHandler(async (req, res, next) => {
+  const { title, description, read } = req.body;
+
+  const notification = await new Notification({
+    title,
+    description,
+    read,
+    userId: req.user.id,
```
Comment (rajivtitus) [inline server/controllers/notification.js:14]: Won't the `userId` here be the recipient's id and not that of the logged in user? :)

File: server/controllers/notification.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    data: unreadNotifications,
+    count: unreadNotifications.length,
+  });
+});
+
+// @desc Mark notifications as read
+// @route PATCH /notifications/read/:notificationId
+// @access Private
+exports.markNotificationAsRead = asyncHandler(async (req, res, next) => {
+  const notification = await Notification.findById(req.params.notificationId);
```
Comment (rajivtitus) [inline server/controllers/notification.js:46]: Optional: You can validate the path parameters and send a `400` if they are missing

**Review by rajivtitus — COMMENTED** [review]
Good work this this Lukman! 
Just a few comments for you to have a look and address.

Note: In the future, please make sure we only have changes pertaining to the feature within a single PR/branch

File: server/controllers/notification.js (hunk 9 above)
Comment (sholabioku) [reply server/controllers/notification.js:14]: @rajivtitus , I think this method is creating notification by logged in user, I seem not to understand where the recipient is coming from. Kindly explain better please.

Comment (rajivtitus) [reply server/controllers/notification.js:14]: @sholabioku You can send the `userId` in the request body as opposed to using the logged in user's id.

Take for example, a message being sent from user A to B. We would need to notify user B that user A has sent them a message. However, when we create a notification when sending a message, the notification will be created for the logged in user (user A) - which is not intended

**Review by rajivtitus — APPROVED** [review]
LGTM, thanks for making those changes 👍🏽

**Pull request merged**